src/main/scala/org/broadinstitute/dsde/firecloud/core/AgoraPermissionHandler.scala

```diff
@@ -0,0 +1,307 @@
+      //any non-empty user is okay!
+      case _ => true
+    }
+  }
+
+
+  def hasValidRole(fireCloudPermission: FireCloudPermission):Boolean = {
+    fireCloudPermission.role match {
+      case Some("NO ACCESS") => true
+      case Some("OWNER") => true
+      case Some("READER") => true
+      //anything that's one of the above-3 is good, anything else is bad
+      case _ => false
+    }
+  }
+
+
+  // translation between a FireCloud role and a list of Agora roles
+  def toAgoraRoles(fireCloudRole:Option[String]) = {
+    fireCloudRole match {
+      case None => List("Nothing")
+      case Some("OWNER") => List("Read","Write","Create","Redact","Manage") // Could use "All" instead but this is more precise
+      case Some("READER") => List("Read")
+      case Some("NO ACCESS") => List("Nothing")
+      case _ => List("Nothing") // TODO: throw an exception instead? Log something?
+    }
+  }
+
+  // translation between a list of Agora roles and a FireCloud role
+  def toFireCloudRole(agoraRoles:Option[List[String]]) = {
+    agoraRoles match {
+      case None => "NO ACCESS"
+      case Some(r) => {
+        r.sorted match {
+          case List("All") => "OWNER"
+          case List("Create","Manage","Read","Redact","Write") => "OWNER"
+          case List("Read") => "READER"
+          case List("Nothing") => "NO ACCESS"
+          case _ => "NO ACCESS" // TODO: throw an exception instead? Log something?
+        }
+      }
+    }
+  }
+}
+
+class AgoraPermissionActor (requestContext: RequestContext) extends Actor with FireCloudRequestBuilding {
+
+  implicit val system = context.system
+  import system.dispatcher
+
+  val log = Logging(system, getClass)
+
+  // TODO: tests
+
+  def receive = {
+    // GET requests
+    case AgoraPermissionHandler.Get(url: String) =>
+      val pipeline = authHeaders(requestContext) ~> sendReceive
+      val permissionListFuture: Future[HttpResponse] = pipeline { Get(url) }
+      permissionListFuture onComplete {
+        case Success(response) =>
+          response.status match {
+            case StatusCodes.OK =>
+              val agoraPermissions = unmarshal[List[AgoraPermission]].apply(response)
+              val fireCloudPermissions = agoraPermissions.map(x => x.toFireCloudPermission)
+              context.parent ! RequestComplete(OK, fireCloudPermissions)
+              context stop self
+            case x =>
+              context.parent ! RequestComplete(x, response.entity)
+              context stop self
+          }
+        case Failure(e) =>
+          context.parent ! RequestComplete(StatusCodes.InternalServerError, e.getMessage)
+          context stop self
+      }
+
+    case AgoraPermissionHandler.Post(url: String, agoraPermissions: List[AgoraPermission]) =>
+      val pipeline = authHeaders(requestContext) ~> sendReceive
+      val permissionListFuture: Future[HttpResponse] = pipeline { Post(url.toString,
+        HttpClient.createJsonHttpEntity(agoraPermissions.toString()))
```

Comment:
I don't like all of this casting/conversion. You should be able to just write this:
```
Post(url, agoraPermissions)
```
and it should just *work*. If it doesn't, then url and agoraPermissions might need extra attention.